sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClient.java

```diff
@@ -1460,6 +1469,141 @@ public Mono<Void> rollbackTransaction(ServiceBusTransactionContext transactionCo
+                // not delete the requested batch size in a single call, even when there are enough messages to do so.
+                // This logic should check "deleted == MAX_DELETE_MESSAGES_COUNT". Until this is fixed, we'll need to
+                // loop if there were any messages deleted, which will cost an extra service call.
+                if (deleted > 0) {
+                    return deleteMessages(MAX_DELETE_MESSAGES_COUNT, deleteMessagesOptions, receiverOptions.getSessionId());
+                } else {
+                    return Mono.empty();
+                }
+            })
+            .reduce(0, Integer::sum);
+    }
+
+    /**
+     * Deletes up to {@code messageCount} messages from the entity. The actual number of deleted messages may be less
+     * if there are fewer eligible messages in the entity.
+     *
+     * <p>If the lock for a message is held by a receiver, it will be respected and the message will not be deleted.</p>
+     *
+     * @param messageCount the desired number of messages to delete.
+     * @param options options used to delete the messages.
+     * @param sessionId Session id of the messages to delete from. {@code null} if there is no session.
+     * @return a {@link Mono} indicating the number of messages deleted.
+     *
+     * @throws IllegalArgumentException when the {@code messageCount} is less than 1 or exceeds the maximum allowed, as
+     * determined by the Service Bus service.
+     */
+    private Mono<Integer> deleteMessages(int messageCount, DeleteMessagesOptions options, String sessionId) {
+        if (isDisposed.get()) {
+            return monoError(LOGGER, new IllegalStateException(
+                String.format(INVALID_OPERATION_DISPOSED_RECEIVER, "deleteMessages")));
+        }
+        if (Objects.isNull(options)) {
+            return monoError(LOGGER, new NullPointerException("'options' cannot be null."));
+        }
+        if (messageCount <= 0 || messageCount > MAX_DELETE_MESSAGES_COUNT) {
+            final String message = "'messageCount' must be a positive number and less than " + MAX_DELETE_MESSAGES_COUNT;
+            return monoError(LOGGER, new IllegalArgumentException(message));
+        }
+        final OffsetDateTime beforeEnqueueTimeUtc;
+        if (options.getBeforeEnqueueTimeUtc() != null) {
+            beforeEnqueueTimeUtc = options.getBeforeEnqueueTimeUtc();
+        } else {
+            beforeEnqueueTimeUtc = OffsetDateTime.now(ZoneOffset.UTC);
+        }
+        final Mono<Integer> deleteMessages = connectionProcessor
+            .flatMap(connection -> connection.getManagementNode(entityPath, entityType))
+            .flatMap(managementNode -> managementNode.deleteMessages(messageCount, beforeEnqueueTimeUtc, getLinkName(sessionId), sessionId));
+
+        return withRetry(deleteMessages, connectionCacheWrapper.getRetryOptions(),
```

Comment:
don't we handle retries in `connection.getManagementNode` and `managementNode.deleteMessages` already?

Reply:
We don't do retry there, actually, I believe we can't do retry in those levels because within "Connection.getManagementNode" and "Connection::ManagementNode::&lt;any-operation&gt;" scope we don't have access to the reactor-chain to get "a new connection" in case that owning connection instance is disposed. A retry within a connection instance scope can only recover a child of that connection instance (session, link) assuming the connection is healthy.

Reply:
Should we add retry mechanism to `getManagementNode` then? We call it a lot (e.g. updateDisposition, sessionState methods, etc) and never retry.

All operations on the RequestResponseChannel that are done with `sendWithAck` are retried underneath though.

Reply:
Yes, I think we should centralize it but should brainstorm a bit more. It would be better to do it as a separate work item considering the scope. There are APIs where management operation is leaf action such as _updateDisposition_, _deleteMessages_. But we’ve other functionalities where management operation is intermediate action such as authorization for obtaining a link, we want to sketch how recovery flow in such cases is going to be impacted. I see this work-item https://github.com/Azure/azure-sdk-for-java/issues/38733 which is related.

Reply:
sounds good, thank you!